### DISCUSSION TITLE
Weird cache behavior with subscription and query

### DISCUSSION THREAD

**Discussion by focux** [urql-graphql/urql#2565] [Q&A] [answered]
I'm not sure if I'm experiencing a bug, but it feels like that. So I'm working on a chat using subscriptions and having some issues when using an updater for the `newMessages` subscriptions. When I get many subscription messages and I try to push those messages into an array using `cache.updateQuery`, the cached result returned by `cache.updateQuery` is stale—at glance, it looks like the cache didn't add to the array the messages it got before but after one day of debugging I find out that the issue was due to a query that I was firing every time I was getting a new message and I honestly don't know why that query is affecting the cached results. 

I simplified my case to something like this:

<details>
<summary>
(1) This is the query that I'm trying to update to add new messages
</summary>

```graphql
 query getConversation(id: $id) {
  conversation(id: $id) {
     # __typename = Conversation
    id
    messages {
        # __typename = ConversationMessage
        id
        text
    }
  }
}
```
</details>
<details>
<summary>
(2) This is the subscription
</summary>

```graphql
 subscription newConversationMessage(conversationId: $conversationId) {
  newConversationMessage(id: $conversationId) {
    message{
         # __typename = ConversationMessage
        id
        text
    }
    conversation {
        # __typename = Conversation
        id
    }
  }
}
```
</details>
<details>
<summary>
(3) This is the query that I'm firing every time I get a new message
</summary>

```graphql
 query markConversationAsRead(conversationId: $conversationId) {
  markConversationAsRead(id: $conversationId) {
     # __typename = Conversation
    id
    unreadMessages
  }
}
```
</details>
<details>
<summary>
(4) This is the subscription updater
</summary>

```typescript
cache.updateQuery(
  {
    query: MessageList,
    variables: {
      input: { id: result.newConversationMessage.conversation.id },
    },
  },
  (data) => {
    data.chatRoom.messages.push(result.newConversationMessage.message);

    return data;
  }
);
```
</details>

So it looks like is some kind of conflict or race condition between the query I fire every time I get a new subscription message (3) and the query I'm trying to update with the subscription updater (1), I think it's related to (3) returning partial data and URQL merging that data with the query I'm updating (1) while the subscription (2) is trying to update.

Any idea whether this is a bug or not? And if it isn't, can someone explain to me how that works and maybe a workaround?

**Comment by kitten** [comment, marked as answer]
Sorry, I can't take a look at the entire question and text here, but I have some quick notes on the various details here.

Re. `query markConversationAsRead`; This _must_ be a mutation, not a query, to become a side-effect. You can then also optimistically update the result, but the main point here is that you're querying and expect results to change. However, the timing here isn't trivial, and it should be a mutation to clarify the timing to Graphcache.

When you have a subscription and query in parallel there is no deterministic order of them. Graphcache will always attempt to do the best case possible to always behave the same. This doesn't mean it can guess what you want to do correctly, but it'll attempt to be internally consistent.

In general, you'll have to decide whether you want to update a live list, like a chat, via subscriptions or queries. When you have the two deliver the "same" updates and you only react to the subscription, your updates may be overwritten by pending queries.

We recently made a change here to always bump subscription updates up, i.e. bump layers in our cache up to the highest possible (non-fulfilled) priority when they're changing.

This means that your subscription will write to its own layer while other queries are ongoing, which will be the most important layer while queries are pending. As the queries complete they will therefore only take priority if they've started before the subscription update has come in. Otherwise, their result and layer will take priority. If you make conflicting updates in the two, then they can't both be applied.

**Reply by focux** [reply]
Thank you for your answer. I made mistake when trying to simplify my case but `markConversationAsRead` is indeed a mutation.

Also, if I got you right, having mutations and subscriptions update the same data will almost always result in conflicts? So I guess I should return no data on `markConversationAsRead` and instead update what needs to be updated through updaters on that mutation?

**Reply by kitten** [reply]
Well, mutations and subscriptions shouldn't conflict. There's obviously the issue of sending a mutation, having an update from the mutation and receiving an update from the subscription prior to the mutation.

In the past I've dealt with this by guaranteeing that the events for the subscriptions cannot outpace the mutation response. It may obviously still happen, due to transport delays, but it makes some conflicts less likely 

Then re. the lists. It's hard for me to tell on the spot what kinds of conflicts you may see. The trouble lies always in updating the same list multiple times. The reason that's problematic is that you may be writing more data to the "layers" that Graphcache keeps than intended 

One way to prevent conflicts here is to always try to do the minimum amount of updates that'll always work. We have a special `cache.link` method to only update a single "link" (i.e. a relationship; a property that carries a single or a list of normalised entities)

You can still use `updateQuery` but I'd then recommend to write a query inline that only basically carries the list you want to update and has only the keyable field for the items (so presumably `id` for the messages)

This way you'll only update the list itself.

However, this still means that you may be updating the list multiple times. To avoid conflicts there you basically want to avoid updating the list with conflicting information whenever possible.
That's easily said obviously, but I hope it makes sense ✌️

**Reply by focux** [reply]
I really appreciate you taking the time to answer my doubts. I was using Apollo before so all the things that make sense to me are because of that background and not really because it's how it must work.

One last question, how do you handle partial updates when using the graph cache, schema awareness, and cache-and-network policy? Let's suppose I have already some data cached for a player entity:

```typescript
# Player:1
{
  id: 1,
  firstName: 'Jon',
  lastName: 'Doe',
}
```

And then I execute another query which returns this data nested on the result:

```typescript
# Player:1
{
  id: 1,
  firstName: 'Jon',
}
```

How do you merge that data with the cached one, well, do you even merge it? If not, do you trigger a fetch request to update that data instead of merging it? If you fetch again to update the entity, do you trigger the fetch immediately? I tried to look into the docs but I didn't find anything related to this.
